hecate_lexer/src/lib.rs

```diff
@@ -1 +1,225 @@
+use crate::token::Operator::*;
+use crate::token::Token;
+use hecate_util::span::{Source, Span, Spanned};
+use std::{
+    iter::{Iterator, Peekable},
+    str::Chars,
+};
+
 pub mod token;
+
+#[cfg(test)]
+mod tests;
+
+pub struct Lexer<'a> {
+    input: &'a str,
+    iter: Peekable<Chars<'a>>,
+    current_pos: usize,
+}
+
+impl<'a> Lexer<'a> {
+    pub fn new(input: &'a str) -> Lexer<'_> {
+        let iter = input.chars().peekable();
+        Lexer {
+            input,
+            iter,
+            current_pos: 0,
+        }
+    }
+
+    fn advance(&mut self) {
+        self.iter.next();
+        self.current_pos += 1;
+    }
+}
+
+impl<'a> Iterator for Lexer<'a> {
+    type Item = Spanned<'a, Token>;
+    fn next(&mut self) -> Option<Self::Item> {
+        // skip all white spaces
+        loop {
+            match self.iter.peek() {
+                None => {
+                    return Option::Some(Spanned {
+                        inner: Token::EOF,
+                        loc: Span::Generated,
```

Comment:
maybe loc should be the _location_ of the last char (instead of generated) so that an error message can accurately point to EOF when this is part of the error:

```rs
fun foo {
        ^
  unexpeted end of file
```

Reply:
should be fixed by `df9b03c`